### PR TITLE
SAK-49486 Plus improve launch logic and use executor pattern for NRPS retrieval

### PR DESCRIPTION
**Pull request by csev**
While investigating several other JIRAs, I did a detailed review and testing of the Plus "first launch" code because I thought those JIRAs were caused by my code.  After all the investigation of those JIRAs it turns out the tutorial was the problem all along so I moved the First launch cleanup to a new JIRA.

But during the review, in hopes of catching whatever error was happening, I strengthened some areas of the first launch code.

Some error messages have been improved, some error messages have been added, some extra checking has been added to catch inconsistencies in new site setup earlier, and the initial roster pull has been moved from a Thread in ProviderServlet to an Executor in the PlusServiceImpl - following the example in RefreshAuthZGroups.

This is the pull request formerly known as SAK-49467 :)

### CODE REVIEW

File: plus/impl/src/main/java/org/sakaiproject/plus/impl/PlusServiceImpl.java [hunk 1]
```diff
@@ -521,11 +567,93 @@ public void connectLinkAndPlacement(Link link, String placementId)
...
+				long start = System.currentTimeMillis();
+
+				try {
+					syncSiteMembershipsInternal(contextGuid);
+				} catch ( LTIException e ) {
+					log.error("refreshContextMembershipsTask.run() Problem refreshing context: " + contextGuid, e);
+				} finally {
+					time = (System.currentTimeMillis() - start);
+					refreshQueue.remove(contextGuid);
+					if (log.isDebugEnabled()) log.debug("refreshContextMembershipsTask.run() refresh of context: " + contextGuid + " took " + time/1e3 + " seconds");
```
Comment (ottenhoff) [inline plus/impl/src/main/java/org/sakaiproject/plus/impl/PlusServiceImpl.java:603]:
```suggestion
					log.debug("refreshContextMembershipsTask.run() refresh of context: {} took seconds", contextGuid, time/1e3);
```

File: plus/impl/src/main/java/org/sakaiproject/plus/impl/PlusServiceImpl.java [hunk 2]
```diff
@@ -521,11 +567,93 @@ public void connectLinkAndPlacement(Link link, String placementId)
...
+
+				timeRefreshed += time;
+				if (time > longestRefreshed) {
+					longestRefreshed = time;
+					longestName = contextGuid;
+				}
+
+			}
+			log.info("refreshContextMembershipsTask.run() refreshed " + numberRefreshed + " contexts in " + timeRefreshed/1e3 +
+				" seconds, longest contexts was " + longestName + " at " + longestRefreshed/1e3 + " seconds");
```
Comment (ern) [inline plus/impl/src/main/java/org/sakaiproject/plus/impl/PlusServiceImpl.java:614]: update the log with the newer pattern

File: plus/impl/src/main/java/org/sakaiproject/plus/impl/PlusServiceImpl.java [hunk 3]
```diff
@@ -521,11 +567,93 @@ public void connectLinkAndPlacement(Link link, String placementId)
...
+				log.debug("Context pulled from queue {}", contextGuid);
+
+				numberRefreshed++;
+				long time = 0;
+				long start = System.currentTimeMillis();
+
+				try {
+					syncSiteMembershipsInternal(contextGuid);
+				} catch ( LTIException e ) {
+					log.error("refreshContextMembershipsTask.run() Problem refreshing context: " + contextGuid, e);
```
Comment (ern) [inline plus/impl/src/main/java/org/sakaiproject/plus/impl/PlusServiceImpl.java:599]: this one too

File: plus/impl/src/main/java/org/sakaiproject/plus/impl/PlusServiceImpl.java [hunk 4]
```diff
@@ -521,11 +567,93 @@ public void connectLinkAndPlacement(Link link, String placementId)
...
+				} finally {
+					time = (System.currentTimeMillis() - start);
+					refreshQueue.remove(contextGuid);
+					log.debug("refreshContextMembershipsTask.run() refresh of context: {} took {} seconds", contextGuid, time/1e3);
```
Comment (ern) [inline plus/impl/src/main/java/org/sakaiproject/plus/impl/PlusServiceImpl.java:603]: this one

**Review by ern — APPROVED** [review]

**Pull request merged**